crates/router/src/core/payments/flows/session_flow.rs

```diff
@@ -945,39 +928,190 @@ fn create_gpay_session_token(
-            ..router_data.clone()
-        })
+                        ..allowed_payment_methods
+                    },
+                )
+                .collect();
+
+            Ok(types::PaymentsSessionRouterData {
+                response: Ok(types::PaymentsResponseData::SessionResponse {
+                    session_token: payment_types::SessionToken::GooglePay(Box::new(
+                        payment_types::GpaySessionTokenResponse::GooglePaySession(
+                            payment_types::GooglePaySessionResponse {
+                                merchant_info: gpay_data.data.merchant_info,
+                                allowed_payment_methods: gpay_allowed_payment_methods,
+                                transaction_info,
+                                connector: connector.connector_name.to_string(),
+                                sdk_next_action: payment_types::SdkNextAction {
+                                    next_action: payment_types::NextActionCall::Confirm,
+                                },
+                                delayed_session_token: false,
+                                secrets: None,
+                                shipping_address_required: required_shipping_contact_fields,
+                                // We pass Email as a required field irrespective of
+                                // collect_billing_details_from_wallet_connector or
+                                // collect_shipping_details_from_wallet_connector as it is common to both.
+                                email_required: required_shipping_contact_fields
+                                    || is_billing_details_required,
+                                shipping_address_parameters:
+                                    api_models::payments::GpayShippingAddressParameters {
+                                        phone_number_required: required_shipping_contact_fields,
+                                    },
+                            },
+                        ),
+                    )),
+                }),
+                ..router_data.clone()
+            })
+        }
     }
 }
 
+/// Card Type for Google Pay Allowerd Payment Methods
+pub(crate) const CARD: &str = "CARD";
+
+fn get_allowed_payment_methods_from_cards(
+    gpay_info: payment_types::GooglePayWalletDetails,
+    gpay_token_specific_data: &payment_types::GooglePayTokenizationSpecification,
+    is_billing_details_required: bool,
+) -> RouterResult<payment_types::GpayAllowedPaymentMethods> {
+    let billing_address_parameters =
+        is_billing_details_required.then_some(payment_types::GpayBillingAddressParameters {
+            phone_number_required: is_billing_details_required,
+            format: payment_types::GpayBillingAddressFormat::FULL,
+        });
+
+    let protocol_version: Option<String> = gpay_token_specific_data
+        .parameters
+        .public_key
+        .clone()
+        .map(|_| PROTOCOL.to_string());
```

Comment:
If you're anyway not using the inner value, you should be able to use `as_ref()` instead of the `clone()`.